salt/modules/xbps-pkg.py

```diff
@@ -0,0 +1,641 @@
+
+
+def _rehash():
+    '''
+    Recomputes internal hash table for the PATH variable.
+    Used whenever a new command is created during the current
+    session.
+    '''
+    shell = __salt__['environ.get']('SHELL')
+    if shell.split('/')[-1] in ('csh', 'tcsh'):
+        __salt__['cmd.run']('rehash', output_loglevel='trace')
+
+
+def __virtual__():
+    '''
+    Set the virtual pkg module if the os is Void and xbps-install found
+    '''
+    if __grains__['os'] in ( 'Void') and _check_xbps():
+        return __virtualname__
+    return False
+
+
+def list_pkgs(versions_as_list=False, **kwargs):
+    '''
+    List the packages currently installed as a dict::
+
+        {'<package_name>': '<version>'}
+
+    .. versionadded:: XXX 201X.XX
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.list_pkgs
+    '''
+    versions_as_list = salt.utils.is_true(versions_as_list)
+    # not yet implemented or not applicable
+    if any([salt.utils.is_true(kwargs.get(x))
+            for x in ('removed', 'purge_desired')]):
+        return {}
+
+    cmd = 'xbps-query -l'
+    ret = {}
+    out = __salt__['cmd.run'](cmd, output_loglevel='trace')
+    for line in out.splitlines():
+        if not line:
+            continue
+        try:
+            # xbps-query -l output sample: 
+            # ii desktop-file-utils-0.22_4  Utilities to ...
+            #
+            # XXX handle package status (like 'ii') ?
+            pkg, ver = line.split(None)[1].rsplit('-', 1)
+        except ValueError:
+            log.error('xbps-query: Unexpected formatting in '
+                      'line: "{0}"'.format(line))
+
+        __salt__['pkg_resource.add_pkg'](ret, pkg, ver)
+    
+    __salt__['pkg_resource.sort_pkglist'](ret)
+    if not versions_as_list:
+        __salt__['pkg_resource.stringify'](ret)
+    return ret
+
+
+def list_upgrades(**kwargs):
+    '''
+    Check whether or not an upgrade is available for all packages
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.list_upgrades
+    '''
+
+    # sample output of 'xbps-install -un':
+    #     fuse-2.9.4_4 update i686 http://repo.voidlinux.eu/current 298133 91688
+    #     xtools-0.34_1 update noarch http://repo.voidlinux.eu/current 21424 10752
+
+    refresh = salt.utils.is_true(kwargs.pop('refresh', True))
+    
+    # Refresh repo index before checking for latest version available
+    if refresh:
+        refresh_db()
+
+    ret = {}
+
+    # retrieve list of updatable packages
+    cmd = 'xbps-install -un'
+    out = __salt__['cmd.run'](cmd, output_loglevel='trace')
+    for line in out.splitlines():
+        if not line:
+            continue
+        try:
+            pkg, ver = line.split()[0].rsplit('-', 1)
+        except (ValueError, IndexError):
+            log.error('xbps-query: Unexpected formatting in '
+                      'line: "{0}"'.format(line))
+            continue
+
+        log.trace('pkg={0} version={1}'.format(pkg, ver))
```

Comment:
If we catch and continue above, `pkg` and `ver` will be unset down here and stacktrace. They may need to be set to defaults above.